File: designpattern/src/main/java/com/seniorlibs/designpattern/ch25v2/report/ConsoleReporter.java
package com.seniorlibs.designpattern.ch25v2.report;

import com.seniorlibs.designpattern.ch25v2.aggregate.Aggregator;
import com.seniorlibs.designpattern.ch25v2.model.RequestInfo;
import com.seniorlibs.designpattern.ch25v2.model.RequestStat;
import com.seniorlibs.designpattern.ch25v2.repository.MetricsStorage;
import com.seniorlibs.designpattern.ch25v2.viewer.StatViewer;

import java.util.List;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * 组装类并定时触发执行统计显示
 */
public class ConsoleReporter {
    private MetricsStorage metricsStorage;
    private Aggregator aggregator;
    private StatViewer viewer;
    private ScheduledExecutorService executor;

    public ConsoleReporter(MetricsStorage metricsStorage, Aggregator aggregator, StatViewer viewer) {
        this.metricsStorage = metricsStorage;
        this.aggregator = aggregator;
        this.viewer = viewer;
        this.executor = Executors.newSingleThreadScheduledExecutor();
    }

    public void startRepeatedReport(long periodInSeconds, long durationInSeconds) {
        executor.scheduleAtFixedRate(new Runnable() {
            @Override
            public void run() {
                long durationInMillis = durationInSeconds * 1000;
                long endTimeInMillis = System.currentTimeMillis();
                long startTimeInMillis = endTimeInMillis - durationInMillis;
                Map<String, List<RequestInfo>> requestInfos =
                        metricsStorage.getRequestInfos(startTimeInMillis, endTimeInMillis);
                Map<String, RequestStat> requestStats = aggregator.aggregate(requestInfos, durationInMillis);
                viewer.output(requestStats, startTimeInMillis, endTimeInMillis);
            }
        }, 0L, periodInSeconds, TimeUnit.SECONDS);
    }
}

File: designpattern/src/main/java/com/seniorlibs/designpattern/principle/frame/ch25v2/report/ConsoleReporter.kt
package com.seniorlibs.designpattern.principle.frame.ch25v2.report

import com.seniorlibs.designpattern.principle.frame.ch25v2.aggregate.Aggregator
import com.seniorlibs.designpattern.principle.frame.ch25v2.repository.MetricsStorage
import com.seniorlibs.designpattern.principle.frame.ch25v2.viewer.StatViewer
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

/**
 * 组装类并定时触发执行统计显示
 */
class ConsoleReporter(
    private val metricsStorage: MetricsStorage,
    private val aggregator: Aggregator,
    private val viewer: StatViewer
) {
    private val executor: ScheduledExecutorService

    fun startRepeatedReport(periodInSeconds: Long, durationInSeconds: Long) {
        executor.scheduleAtFixedRate({
            val durationInMillis = durationInSeconds * 1000
            val endTimeInMillis = System.currentTimeMillis()
            val startTimeInMillis = endTimeInMillis - durationInMillis
            val requestInfos = metricsStorage.getRequestInfos(startTimeInMillis, endTimeInMillis)
            val requestStats = aggregator.aggregate(requestInfos, durationInMillis)
            viewer.output(requestStats, startTimeInMillis, endTimeInMillis)
        }, 0L, periodInSeconds, TimeUnit.SECONDS)
    }

    init {
        executor = Executors.newSingleThreadScheduledExecutor()
    }
}